src/main/java/Backend.java

```diff
@@ -1,149 +1,107 @@
-                            emptypayload[0] = new_key;
-                            emptypayload[1] = new OtpErlangBinary(crdt_object);
-                            OtpErlangTuple new_snapshot = new OtpErlangTuple(emptypayload);
-                            myOtpMbox.send(last_pid, new_snapshot);
-                    }
+                CRDT crdt_object = ObjectTable.get(JavaObjectId).object;
+                GenericFunction func = (GenericFunction) binary.getObject();
+                HashSet<Integer> FunctionIdList = ObjectTable.get(JavaObjectId).idlist;
+                if (!FunctionIdList.contains(func.getId())) {
+                    crdt_object.invoke(func);
+                    FunctionIdList.add(func.getId());
                 }
-            } catch (Exception e) {
-                e.printStackTrace();
-                // return some failure
-                OtpErlangString errormsg = new OtpErlangString("hi");
-                // myOtpMbox.send(last_pid, errormsg);
-                OtpErlangAtom atom = new OtpErlangAtom("error");
-                myOtpMbox.send(last_pid, atom);
+            } catch (ClassCastException e) {
+                /* intentionally ignore this exception */
+                // This happens because we don't have an Idset for crdt initializations through
+                // updates so we got a redundant crdt object(something we already have)
+                assert (binary.getObject().getClass() == CRDT.class);
             }
         }
+
+        return JavaObjectId;
     }
 
-    // The methods below are used in testing
+    public OtpErlangBinary downstream(OtpErlangBinary JavaObjectId, OtpErlangBinary binary) {
+        // TBD what if anything we want to do here otherwise something like this
+        return binary;
+    }
+
+    public OtpErlangTuple snapshot(OtpErlangBinary JavaObjectId) throws NoSuchObjectException {
+        // assert that object is in table else request it
+        CRDT crdt_object = ObjectTable.get(JavaObjectId).object;
+        if (crdt_object == null) {
+            throw new NoSuchObjectException();
+        }
+
+        // ObjectTable.get(JavaObjectId).idlist.clear();
+        // ObjectTable.remove(ERLObjectId);
 
+        OtpErlangBinary new_key = newJavaObjectId();
+        CRDTMapEntry<CRDT, HashSet<Integer>> mappair = new CRDTMapEntry<CRDT, HashSet<Integer>>(
+                crdt_object.deepClone(), new HashSet<Integer>());
+        ObjectTable.put(new_key, mappair);
+        OtpErlangObject[] emptypayload = new OtpErlangObject[2];
+        emptypayload[0] = JavaObjectId; // new_key;
+        emptypayload[1] = new OtpErlangBinary(crdt_object);
+        OtpErlangTuple new_snapshot = new OtpErlangTuple(emptypayload);
+        return new_snapshot;
+    }
+
+    public OtpErlangBinary newJavaObjectId() {
+        byte[] b = new byte[20];
+        new Random().nextBytes(b);
+        return new OtpErlangBinary(b);
+    }
+
+    // The methods below are used in testing
```

Comment:
Protip: You can extend backend to a new class `BackendWithTesting` and stick this stuff there, or you can leave them in place, or you can delete the giant comment block below this comment.  I don't recommend just leaving a bunch of live code in place while keeping it commented out.